### ISSUE TITLE
tools/flow: reference to parent struct does not create dependency on contained tasks

### ISSUE THREAD

**Issue by kghenderson** [cue-lang/cue#2078]
### What version of CUE are you using (`cue version`)?

<pre>
cue version v0.5.0-alpha.1

       -compiler gc
     CGO_ENABLED 1
          GOARCH amd64
            GOOS darwin
         GOAMD64 v1

</pre>

### Does this issue reproduce with the latest release?

if this alpha is considered latest release, then yes

### What did you do?

<pre>
package servetest

import (
	"tool/cli"
)

#Things: [mapkey=string]: #Thing & {Name: mapkey}
#Thing: {Name: string, Purpose: string}

Things: #Things & {
	"MyThing1": {
		Purpose: "To serve butter ..."
	}, "MyThing2": {
		Purpose: "Ketchup bot."
	}, "MyThing3": {
		Purpose: "Aggressively collect crumbs."
	}
}

command: "servetest": {
	doServe: {
		for thingName, thing in Things {
			let serveText = thing.Name + "'s purpose is to " + thing.Purpose
			"doPrint-\(thingName)": cli.Print & {
				text: serveText
			}
		}
	}

	afterServing: {
		// note: careful about the success, otherwise the whole command (will repeat)
		$after: doServe // runs many times, doesn't wait
		//		$after:    doServe.$success  // runs once, but doesn't work
		printDone: cli.Print & {text: "done serving "}
	}
}

</pre>

run: `cue cmd servetest`  
and run it multiple times, since it will sometimes wait

### What did you expect to see?

I expect "afterServing" to wait until doServe has completed, and print only once. 

### What did you see instead?
There are 2 scenarios here: 
1: 
$after: doServe will repeat the doServe N=3 (in collection) times

2: 
$after: doServe.$success, runs once, but does not wait until completion. 

---
note that there is a condensed version of: 
https://github.com/kghenderson/cue-companion/blob/main/examples/direnv_example/tools/serve/serve_tool.cue

**Comment by verdverm** [comment]
Both versions of `$after: <...>` do not create an inter task dependency. You have to refer to an incomplete field within the task(s) you wish to wait on. For your example, you would want something like

```
$after: [ for thingName, _ in Things doServe["doPrint-\(thingName)"].$done ]
```

the subfield of the task to wait on doesn't matter so much, just that it's incomplete. It should be referenceable, so you would need to add `$done: _` in the print tasks, and you never need to fill it. For tasks which have incomplete fields they fill, you can wait on those.

Basically, what is happening in the dependency detection is...
1. does this task reference another field?
2. if it's already concrete, no dependency is needed
3. if it's incomplete, is it within another task?
4. if yes, mark the dependency
5. this task can run when all dependent tasks are completed, it does not recheck (afaik) that the referenced field is actually concrete when it starts the task waiting on prior results

**Comment by kghenderson** [comment]
thank you @verdverm. i'm going to leave this issue open though because i think this behaviour as it is, is very unintuitive. i'd have a hard time trying to explain this to others.

**Comment by myitcv** [comment]
@kghenderson thanks for raising this. My analysis on the problem is slightly different to @verdverm's.

First to point out one thing: the use of `$after` must be within a task. All it does is create a dependency between tasks where there would otherwise not be a reference to create such a dependency. 

In your original code, the `$after` field appears as a sibling of a task.

Consider a slight rewrite of your example therefore:

```cue
package servetest

import (
	"tool/cli"
)

#Things: [mapkey=string]: #Thing & {Name: mapkey}
#Thing: {Name: string, Purpose: string}

Things: #Things & {
	"MyThing1": {
		Purpose: "To serve butter ..."
	}
	"MyThing2": {
		Purpose: "Ketchup bot."
	}
	"MyThing3": {
		Purpose: "Aggressively collect crumbs."
	}
}

command: "servetest": {
	doServe: {
		for thingName, thing in Things {
			let serveText = thing.Name + "'s purpose is to " + thing.Purpose
			(thingName): cli.Print & {
				text: serveText
			}
		}
	}
	printDone: cli.Print & {
		$after: doServe
		text:   "done serving "
	}
}
```

The `$after` field in `printDone` now references the `doServe` struct, which contains the three tasks `MyThingN`. 

Looking at a mermaid dependency graph of the initial dependencies:

```mermaid
graph TD
  t0("command.servetest.doServe.MyThing1 [Ready]")
  t1("command.servetest.doServe.MyThing2 [Ready]")
  t2("command.servetest.doServe.MyThing3 [Ready]")
  t3("command.servetest.printDone [Ready]")
```

This highlights the problem: there is no dependency between any of the `MyThingN` tasks and `printDone`. Consider a further rewrite:

```
package servetest

import (
	"tool/cli"
)

#Things: [mapkey=string]: #Thing & {Name: mapkey}
#Thing: {Name: string, Purpose: string}

Things: #Things & {
	"MyThing1": {
		Purpose: "To serve butter ..."
	}
	"MyThing2": {
		Purpose: "Ketchup bot."
	}
	"MyThing3": {
		Purpose: "Aggressively collect crumbs."
	}
}

command: "servetest": {
	doServe: {
		for thingName, thing in Things {
			let serveText = thing.Name + "'s purpose is to " + thing.Purpose
			(thingName): cli.Print & {
				text: serveText
			}
		}
	}
	printDone: cli.Print & {
		$after: [ for _, v in doServe {v}]
		text: "done serving "
	}
}
```

The change here being in the `$after` field. Now the dependency graph looks like this:

```mermaid
graph TD
  t0("command.servetest.doServe.MyThing1 [Ready]")
  t1("command.servetest.doServe.MyThing2 [Ready]")
  t2("command.servetest.doServe.MyThing3 [Ready]")
  t3("command.servetest.printDone [Waiting]")
  t3-->t0
  t3-->t1
  t3-->t2
```

(note there is no order between the `MyThingN` tasks, I'm assuming that was intended from your side).

So the bug here is that the reference to `doServe` does not capture the three `MyThingN` tasks when it should. That is a bug.

@verdverm note that you don't need to reference a field in a task, a reference to the task itself is sufficient. 

So @kghenderson your intuition here is correct. The problem is a bug is getting in the way! I will repurpose this issue and summarise the bug in the next comment.

**Comment by myitcv** [comment]
Precise summary of the bug described in https://github.com/cue-lang/cue/issues/2078#issuecomment-1315085636.

### What did you do?

```
exec cue cmd servetest
cmp stdout stdout.golden

-- x_tool.cue --
package servetest

import (
	"tool/cli"
)

command: "servetest": {
	doServe: {
		print: cli.Print & {
			text: "this is a test"
		}
	}
	printDone: cli.Print & {
		$after: doServe
		text:   "done serving "
	}
}
-- stdout.golden --
this is a test
done serving
```

### What did you expect to see?

Passing test, 100% of the time.

### What did you see instead?

Failing test, ~50% of the time:

```
$ ts repro.txtar
> exec cue cmd servetest
[stdout]
done serving
this is a test
> cmp stdout stdout.golden
--- stdout
+++ stdout.golden
@@ -1,2 +1,2 @@
-done serving
 this is a test
+done serving

FAIL: /tmp/testscript1603975044/repro.txtar/script.txtar:2: stdout and stdout.golden differ
```

**Comment by myitcv** [comment]
FYI I just landed https://review.gerrithub.io/c/cue-lang/cue/+/546541 to make debugging dependency issues easier.

**Comment by kghenderson** [comment]
i didn't actually realize what a "Task" actually was .. 

@verdverm  made a comment in slack last week that sparked a light-bulb for me, when he said:
> tasks are values ... which have the $id field within them 
> basically, the engine walks the value tree, looking for the $id field

once this dawned on me, i understood the error of my ways (thank you @verdverm  and @myitcv and @rogpeppe ) 
i had simply (incorrectly) assumed that 'parent tasks' were groupings of 'sub-tasks', which obviously isn't the case. 

note that i do think that we should revisit this behaviour, but there's a lot more to unpack with that,
as it ties in to multiple other requests and proposals.

re: referring to a task itself and not $done,  i will retest in light of that potential bug-fix, awesome news, thanks! :-)

**Comment by kghenderson** [comment]
i'll also add that the cue documentation seems to _imply_ (in my flawed interpretation) that tasks are the keys within a command map, i.e. one level down. but i was using the cuetorials commands as reference which showed that wasn't strictly the case, and in my own experiments, the nested "tasks" worked as well. so then i just figured any descendant key of the command was a 'task'. again, this obviously isn't correct, but just wanted to explain the rationale. so this might need to be addressed in the description of 'tasks' on the cue-lang site.

**Comment by myitcv** [comment]
> i had simply (incorrectly) assumed that 'parent tasks' were groupings of 'sub-tasks', which obviously isn't the case.

Just to clarify though, the declaration of a dependency from one task to some other tasks _should_ work when the target is a reference to a value that contains tasks, i.e. a parent struct (because for `cue cmd` structs are walked to find struct values which define an `$id` field, as you point out). That does not work today, but it should. So your intuition is correct.

**Comment by myitcv** [comment]
Adding a note to self: need to clarify whether this is a feature request or not. From memory, the reason I thought this was a bug is that there is/was a place in the CUE code base that suggested this should work. Just noting in passing because we are trying to be brief in our big issue garden!

**Comment by networkhermit** [comment]
I think this is a bug. It's counterintuitive that reference to parent task creates no dependency. For now the workaround `[ for _, v in tasks {v}]` can be used.

It's worth noting that the schema for `$after` is `Task | [...Task]`, not `Tasks`.

https://github.com/cue-lang/cue/blob/72ba5289e517063225172d3ee2b0866df286edb8/pkg/tool/tool.cue#L58-L76

**Comment by verdverm** [comment]
The way cue/flow determines dependencies, any field within a task referencing another task will (should) create a dependency. It's basically a big walk over the CUE value / lattice at the ADT level to find any references.

I believe the `$after` was more to let people know by having something in the schema they see in the docs.
